pkg/syncer/storage_policy_quota_reconciler.go

```diff
@@ -0,0 +1,224 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	cnsoperatorv1alpha1 "sigs.k8s.io/vsphere-csi-driver/v3/pkg/apis/cnsoperator"
+	storagepolicyusagev1alpha1 "sigs.k8s.io/vsphere-csi-driver/v3/pkg/apis/cnsoperator/storagepolicy/v1alpha1"
+	"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/v3/pkg/kubernetes"
+)
+
+const (
+	// Annotation added locally on StoragePolicyQuota CR that is being deleted
+	deletionAnnKey = "cns.vmware.com/is-storagepolicyquota-being-deleted"
+	deletionAnnVal = "true"
+)
+
+// StoragePolicyQuotaReconciler is the interface for StoragePolicyQuota reconciler.
+type StoragePolicyQuotaReconciler interface {
+	// Run starts the reconciler.
+	Run(ctx context.Context, workers int)
+}
+
+type storagePolicyQuotaReconciler struct {
+	// metadataSyncer informer to get FSS information.
+	metadataSyncerInformer *metadataSyncInformer
+	// StoragePolicyQuota queue to add/delete StoragePolicyUsage CRs.
+	svcQuotaOpsQueue workqueue.RateLimitingInterface
+	// StoragePolicyQuota Synced.
+	svcQuotaSynced cache.InformerSynced
+}
+
+// newStoragePolicyQuotaReconciler returns a StoragePolicyQuotaReconciler.
+func newStoragePolicyQuotaReconciler(
+	ctx context.Context,
+	metadataSyncerInformer *metadataSyncInformer,
+	svcStoragePolicyQuotaRateLimiter workqueue.RateLimiter,
+	stopCh <-chan struct{}) (*storagePolicyQuotaReconciler, error) {
+
+	log := logger.GetLogger(ctx)
+	// Create an informer for StoragePolicyQuota instances.
+	k8sConfig, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "newStoragePolicyQuotaReconciler: failed to get kubeconfig with error: %v", err)
+	}
+	dynamicInformerFactory, err := k8s.NewDynamicInformerFactory(ctx, k8sConfig, metav1.NamespaceAll, true)
+	if err != nil {
+		log.Errorf("newStoragePolicyQuotaReconciler: could not retrieve dynamic informer factory. Error: %+v", err)
+		return nil, err
+	}
+	// Return informer from shared dynamic informer factory for input resource.
+	gvr := schema.GroupVersionResource{Group: cnsoperatorv1alpha1.GroupName, Version: cnsoperatorv1alpha1.Version,
+		Resource: cnsoperatorv1alpha1.CnsStoragePolicyQuotaPlural}
+	storagePolicyQuotaInformer := dynamicInformerFactory.ForResource(gvr)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "newStoragePolicyQuotaReconciler: failed to get informer with error: %v", err)
+	}
+	svcQuotaOpsQueue := workqueue.NewNamedRateLimitingQueue(
```

Comment:
NewNamedRateLimitingQueue is deprecated, should use NewRateLimitingQueueWithConfig

Reply:
done